### PR TITLE
[SPARK-26998][CORE] Remove SSL configuration from executors

### PR DESCRIPTION
**Pull request by gaborgsomogyi**
## What changes were proposed in this pull request?

Different SSL passwords shown up as command line argument on executor side in standalone mode:
* keyStorePassword
* keyPassword
* trustStorePassword

In this PR I've removed SSL configurations from executors.

## How was this patch tested?

Existing + additional unit tests.
Additionally tested with standalone mode and checked the command line arguments:
```
[gaborsomogyi:~/spark] SPARK-26998(+4/-0,3)+ ± jps
94803 CoarseGrainedExecutorBackend
94818 Jps
90149 RemoteMavenServer
91925 Nailgun
94793 SparkSubmit
94680 Worker
94556 Master
398
[gaborsomogyi:~/spark] SPARK-26998(+4/-1,3)+ ± ps -ef | egrep "94556|94680|94793|94803"
  502 94556     1   0  2:02PM ttys007    0:07.39 /Library/Java/JavaVirtualMachines/jdk1.8.0_152.jdk/Contents/Home/bin/java -cp /Users/gaborsomogyi/spark/conf/:/Users/gaborsomogyi/spark/assembly/target/scala-2.12/jars/* -Xmx1g org.apache.spark.deploy.master.Master --host gsomogyi-MBP.local --port 7077 --webui-port 8080 --properties-file conf/spark-defaults.conf
  502 94680     1   0  2:02PM ttys007    0:07.27 /Library/Java/JavaVirtualMachines/jdk1.8.0_152.jdk/Contents/Home/bin/java -cp /Users/gaborsomogyi/spark/conf/:/Users/gaborsomogyi/spark/assembly/target/scala-2.12/jars/* -Xmx1g org.apache.spark.deploy.worker.Worker --webui-port 8081 --properties-file conf/spark-defaults.conf spark://gsomogyi-MBP.local:7077
  502 94793 94782   0  2:02PM ttys007    0:35.52 /Library/Java/JavaVirtualMachines/jdk1.8.0_152.jdk/Contents/Home/bin/java -cp /Users/gaborsomogyi/spark/conf/:/Users/gaborsomogyi/spark/assembly/target/scala-2.12/jars/* -Dscala.usejavacp=true -Xmx1g org.apache.spark.deploy.SparkSubmit --master spark://gsomogyi-MBP.local:7077 --class org.apache.spark.repl.Main --name Spark shell spark-shell
  502 94803 94680   0  2:03PM ttys007    0:05.20 /Library/Java/JavaVirtualMachines/jdk1.8.0_152.jdk/Contents/Home/bin/java -cp /Users/gaborsomogyi/spark/conf/:/Users/gaborsomogyi/spark/assembly/target/scala-2.12/jars/* -Xmx1024M -Dspark.ssl.ui.port=0 -Dspark.driver.port=60902 org.apache.spark.executor.CoarseGrainedExecutorBackend --driver-url spark://CoarseGrainedScheduler@172.30.65.186:60902 --executor-id 0 --hostname 172.30.65.186 --cores 8 --app-id app-20190326140311-0000 --worker-url spark://Worker@172.30.65.186:60899
  502 94910 57352   0  2:05PM ttys008    0:00.00 egrep 94556|94680|94793|94803
```

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/SSLOptions.scala [hunk 1]
```diff
@@ -131,8 +132,9 @@ private[spark] case class SSLOptions(
   /** Returns a string representation of this SSLOptions with all the passwords masked. */
   override def toString: String = s"SSLOptions{enabled=$enabled, port=$port, " +
       s"keyStore=$keyStore, keyStorePassword=${keyStorePassword.map(_ => "xxx")}, " +
+      s"keyPassword=${keyPassword.map(_ => "xxx")}, keyStoreType=$keyStoreType, " +
```
Comment (HeartSaVioR) [inline core/src/main/scala/org/apache/spark/SSLOptions.scala]: nit: we could add `needClientAuth` as well if we would like to dump all.

File: core/src/main/scala/org/apache/spark/SSLOptions.scala [hunk 2]
```diff
@@ -231,5 +230,45 @@ private[spark] object SSLOptions extends Logging {
...
+  private def getPassword(
+      conf: SparkConf,
+      hadoopConf: Configuration,
+      ns: String,
+      parameter: String,
+      default: Option[String]): Option[String] = {
+    var parameterValue = conf.getWithSubstitution(s"$ns.$parameter")
+      .orElse(Option(hadoopConf.getPassword(s"$ns.$parameter")).map(new String(_)))
+      .orElse(default)
+    if (parameterValue.isDefined && default.isDefined && parameterValue.get != default.get) {
```
Comment (HeartSaVioR) [inline core/src/main/scala/org/apache/spark/SSLOptions.scala]: I have no idea so I'm asking, Is there a case which default can be `None`? If then I guess we also need to warn if `(parameterValue.isDefined && !default.isDefined)` is true. 

One liner (but long) would be `if (parameterValue.isDefined && (!default.isDefined || (parameterValue.get != default.get)))`

File: core/src/main/scala/org/apache/spark/SSLOptions.scala [hunk 3]
```diff
@@ -231,5 +230,45 @@ private[spark] object SSLOptions extends Logging {
...
+      )
+    }
+
+    val parameterFileValue = conf.getWithSubstitution(s"$ns.${parameter}File")
+      .orElse(Option(hadoopConf.getPassword(s"$ns.${parameter}File")).map(new String(_)))
+    if (parameterFileValue.isDefined) {
+      val parameterFileContent = readPasswordFile(parameterFileValue.get)
+      if (parameterValue.isDefined && parameterValue.get != parameterFileContent) {
+        throw new IllegalArgumentException(s"Both $ns.$parameter and $ns.${parameter}File " +
+          "parameters defined but the they differ.")
```
Comment (HeartSaVioR) [inline core/src/main/scala/org/apache/spark/SSLOptions.scala]: nit: remove `the`

**Review by HeartSaVioR — COMMENTED** [review]
Looks good overall. Minor and nit comments.

File: core/src/main/scala/org/apache/spark/SSLOptions.scala (hunk 1 above)
Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/SSLOptions.scala]: Yeah, added to make it complete.

File: core/src/main/scala/org/apache/spark/SSLOptions.scala (hunk 2 above)
Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/SSLOptions.scala]: It's a valid possibility so changed.

File: core/src/main/scala/org/apache/spark/SSLOptions.scala (hunk 3 above)
Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/SSLOptions.scala]: Removed.

**Comment by gaborgsomogyi** [comment]
cc @vanzin

**Comment by gaborgsomogyi** [comment]
Seems like there is a jenkins issue...

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by vanzin** [comment]
So there are two things here.

First, the bug. It's about the value showing up in the command line of executors. There was a time when executors might need this SSL configuration, but they don't anymore; so the SSL-related things can just be removed from the list (see `SparkConf.isExecutorStartupConf`).

Second, even if the above were not true, I think this wouldn't be the right fix. I'd instead have written the stuff returned by `SparkConf.isExecutorStartupConf` to a file, and have the executor process pick up the config from there.

But given the first, I think just can be a one line fix in `SparkConf` instead.

**Comment by gaborgsomogyi** [comment]
> There was a time when executors might need this SSL configuration, but they don't anymore

Thanks for helping me out.

The bug is about the executor side but the same problem exists when one provides config like `spark-shell --conf spark.ssl.keyPassword=foo`.

```
  502 80304 80292   0  9:22AM ttys008    0:34.34 /Library/Java/JavaVirtualMachines/jdk1.8.0_152.jdk/Contents/Home/bin/java -cp /usr/local/Cellar/apache-spark/2.4.0/libexec/conf/:/usr/local/Cellar/apache-spark/2.4.0/libexec/jars/* -Dscala.usejavacp=true -Xmx1g org.apache.spark.deploy.SparkSubmit --conf spark.ssl.keyPassword=foo --class org.apache.spark.repl.Main --name Spark shell spark-shell
```

All in all I've created this PR to solve this as a whole. Do you think the driver side shouldn't have to be solved or just not in this PR?

**Comment by vanzin** [comment]
In that case you can already provide a config file with `--properties-file`, so I don't see the point in adding yet another config file. And properties also can refer to env variables, for example, so you don't even need to hardcode the password in the file itself.

`--conf spark.ssl.keyPassword='${env:KEYSTORE_PASSWD}'`

**Comment by gaborgsomogyi** [comment]
The trick is I've tried originally `--conf spark.ssl.keyPassword="$KEYSTORE_PASSWD"` which is evaluated at shell script level. `'${env:KEYSTORE_PASSWD}'` is not highlighting the password.

File: core/src/test/scala/org/apache/spark/SparkConfSuite.scala [hunk 4]
```diff
@@ -354,6 +354,20 @@ class SparkConfSuite extends SparkFunSuite with LocalSparkContext with ResetSyst
     }
   }
 
+  test("SPARK-26998: SSL configuration not needed on executors") {
+    val conf = new SparkConf(false)
+    conf.validateSettings()
```
Comment (attilapiros) [inline core/src/test/scala/org/apache/spark/SparkConfSuite.scala]: This line can be removed as `validateSettings()` is enough to be called once in line 365.

**Review by attilapiros — COMMENTED** [review]
one minor thing otherwise LGTM

**Comment by vanzin** [comment]
> '${env:KEY_PASSWD}' is not highlighting the password.

I don't know what that means. Here's a line from how we deploy our SHSs:

```
# cat 666-spark_on_yarn-SPARK_YARN_HISTORY_SERVER/spark-conf/spark-history-server.conf | grep keyStorePass
spark.ssl.historyServer.keyStorePassword=${env:KEYSTORE_PASSWORD}
```

**Comment by vanzin** [comment]
(That being said the latest patch looks good.)

**Comment by gaborgsomogyi** [comment]
The reason why I've changed the approach is because I see your point is good.

>'${env:KEY_PASSWD}' is not highlighting the password.

This means not showing the password in plain text but formed the sentence in a bad way.

File: core/src/test/scala/org/apache/spark/SparkConfSuite.scala (hunk 4 above)
Comment (vanzin) [reply core/src/test/scala/org/apache/spark/SparkConfSuite.scala]: Ping. (In fact do you need to call this method at all in this test.)

File: core/src/test/scala/org/apache/spark/SparkConfSuite.scala [hunk 5]
```diff
@@ -354,6 +354,20 @@ class SparkConfSuite extends SparkFunSuite with LocalSparkContext with ResetSyst
...
+
+    conf.set("spark.ssl.enabled", "true")
+    conf.set("spark.ssl.keyPassword", "password")
+    conf.set("spark.ssl.keyStorePassword", "password")
+    conf.set("spark.ssl.trustStorePassword", "password")
+    conf.validateSettings()
+
+    val filtered = conf.getAll.filter { case (k, _) => SparkConf.isExecutorStartupConf(k) }
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/SparkConfSuite.scala]: To make the assert more fine-grained, you could `assert` that each isn't an executor conf with `foreach`

Comment (gaborgsomogyi) [reply core/src/test/scala/org/apache/spark/SparkConfSuite.scala]: The exception tells exactly what is the problematic parameter and value so not sure a more complex construct would help here:
```
Array((spark.network.foo,bar)) was not empty
ScalaTestFailureLocation: org.apache.spark.SparkConfSuite at (SparkConfSuite.scala:366)
org.scalatest.exceptions.TestFailedException: Array((spark.network.foo,bar)) was not empty
```

File: core/src/test/scala/org/apache/spark/SparkConfSuite.scala (hunk 4 above)
Comment (gaborgsomogyi) [reply core/src/test/scala/org/apache/spark/SparkConfSuite.scala]: Yep, I'm dealing with multi-cluster Kafka delegation token. Thanks for pinging me, forgot this but now fixed.

**Review by attilapiros — COMMENTED** [review]
LGTM

File: core/src/test/scala/org/apache/spark/SparkConfSuite.scala (hunk 5 above)
Comment (srowen) [reply core/src/test/scala/org/apache/spark/SparkConfSuite.scala]: OK that's fine

**Comment by vanzin** [comment]
Merging to master / 2.4 / 2.3.

**Pull request closed without merging**